rollbar-java/src/main/java/com/rollbar/notifier/sender/BufferedSender.java

```diff
@@ -0,0 +1,206 @@
+
+import com.rollbar.api.payload.Payload;
+import com.rollbar.notifier.config.Config;
+import com.rollbar.notifier.sender.exception.SenderException;
+import com.rollbar.notifier.sender.listener.SenderListener;
+import com.rollbar.notifier.sender.queue.DiskQueue;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A buffered sender implementation.
+ */
+public class BufferedSender implements Sender {
+
+  private static final int DEFAULT_BATCH_SIZE = Integer.MAX_VALUE;
+
+  private static final long DEFAULT_FLUSH_FREQ = TimeUnit.SECONDS.toMillis(5);
+
+  private final int batchSize;
+
+  private Sender sender;
+
+  private Queue<Payload> queue;
+
+  private ScheduledExecutorService executorService;
+
+  BufferedSender(Builder builder) {
+    this(builder, Executors.newSingleThreadScheduledExecutor(new SenderThreadFactory()));
+  }
+
+  BufferedSender(Builder builder, ScheduledExecutorService executorService) {
+    Objects.requireNonNull(builder.sender, "The sender can not be null");
+    Objects.requireNonNull(builder.queue, "The queue can not be null");
+
+    this.batchSize = builder.batchSize;
+    this.sender = builder.sender;
+    this.queue = builder.queue;
+
+    // Schedule executor service to send events in background with a thread factory that sets the
+    // thread as daemons to allow the jvm exit.
+    this.executorService =  executorService;
+    this.executorService.scheduleWithFixedDelay(new SendTask(batchSize, queue, sender),
+        builder.flushFreq, builder.flushFreq, TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void send(Payload payload) {
+    try {
+      // If the queue is full it will raise an exception and it will be notified.
+      queue.add(payload);
+    } catch (Exception e) {
+      notifyError(payload, new SenderException(e));
+    }
+  }
+
+  @Override
+  public void addListener(SenderListener listener) {
+    sender.addListener(listener);
+  }
+
+  @Override
+  public List<SenderListener> getListeners() {
+    return sender.getListeners();
+  }
+
+  @Override
+  public void close() throws Exception {
+    this.executorService.shutdown();
+    this.sender.close();
+  }
+
+  private void notifyError(Payload payload, Exception e) {
+    for (SenderListener listener : sender.getListeners()) {
+      listener.onError(payload, e);
+    }
+  }
+
+  /**
+   * Builder class for {@link BufferedSender}.
+   */
+  public static final class Builder {
+
+    private int batchSize;
+
+    private long flushFreq;
+
+    private Queue<Payload> queue;
+
+    private Sender sender;
+
+    /**
+     * Constructor.
+     */
+    public Builder() {
+      this.batchSize = DEFAULT_BATCH_SIZE;
+      this.flushFreq = DEFAULT_FLUSH_FREQ;
+      this.queue = new DiskQueue.Builder().build();
```

Comment:
This BufferedSender implementation allows any `Queue<Payload>` to be passed for buffering the events on it. So, this default could be changed by other in memory implementation if you think this should be the correct default one. 

For Android, the current `DiskQueue` implementation should work meanwhile we pass to it the cache directory. It tries to create the folder, only if it doesn't exist, and check some permissions to verify that we can read/write from it. So, I guess that if we pass for Android a folder that already exists and that we can read/write from it, there shouldn't be any problem, but maybe I'm missing something in regards to Android.

Reply:
`new DiskQueue.Builder()` will raise a RuntimeException as it currently is written because of the defaults `new File(".rollbar-queue").mkdirs()` will not work. My point is that I want this constructor to not set the queue. Then I can call `Builder().queue(someAndroidConfiguredDiskQueue).build()` without worrying about what `DiskQueue.Builder().build()` tries to do by default.

Then, in `build` we can do something like

```
public BufferedSender build() {
  if (!this.queue) {
    this.queue = new DiskQueue.Builder().build();
  }
  return new BufferedSender(this);
}
```

Reply:
Oh! yes totally agree. What do you think if we use an in memory queue like ArrayBlockingQueue?